### PR TITLE
fix/top_cities_has_cheapest_price_of_apartments_and_year

### PR DESCRIPTION
**Pull request by khaledeid1k**
ameer don't touch my code ,

### CODE REVIEW

File: src/test/kotlin/interactor/GetTopCitiesHasCheapestPriceOfApartmentsAndCountOfYearToBuyTest.kt [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    @Test
+    fun `should return list when the input salary`() {
+
+        // when the output list of a pair of cities names and count of years
+        val result =
+            getTopCitiesHasCheapestPriceOfApartmentsAndYearInteractor(10)
+
+        // then check
+        assertEquals(
+            listOf(
```
Comment (AmeerAmjed) [inline src/test/kotlin/interactor/GetTopCitiesHasCheapestPriceOfApartmentsAndCountOfYearToBuyTest.kt]: put expected value  in variable

File: src/main/kotlin/interactor/GetTopCitiesHasCheapestPriceOfApartmentsAndYearInteractor.kt [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+            .filter(::excludeNullNetSalaryAndPricePerSquareMeterAndLowDataQuality)
+            .takeIf { it.isNotEmpty() && limit > 0 }
+            ?.run {
+                asSequence()
+                    .distinctBy { it.cityName }
+                    .map(::makeAPairOfCityNameAndCountOfYears)
+                    .sortedBy { it.second }
+                    .take(limit).toList()
+            }
+            ?: throw Exception("Invalid Data")
```
Comment (AmeerAmjed) [inline src/main/kotlin/interactor/GetTopCitiesHasCheapestPriceOfApartmentsAndYearInteractor.kt]: The message is not clear "Invalid Data"

**Review by AmeerAmjed — CHANGES_REQUESTED** [review]

File: src/main/kotlin/interactor/GetTopCitiesHasCheapestPriceOfApartmentsAndYearInteractor.kt (hunk 2 above)
Comment (khaledeid1k) [reply src/main/kotlin/interactor/GetTopCitiesHasCheapestPriceOfApartmentsAndYearInteractor.kt]: what your suggestion to name it

File: src/test/kotlin/interactor/GetTopCitiesHasCheapestPriceOfApartmentsAndCountOfYearToBuyTest.kt [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    private lateinit var getTopCitiesHasCheapestPriceOfApartmentsAndYearInteractor:
+            GetTopCitiesHasCheapestPriceOfApartmentsAndYearInteractor
+    private lateinit var fakeDataSource: CheapestApartmentsInCitiesFakeData
+    private lateinit var emptyDataSource: EmptyFakeData
+    private lateinit var invalidDataSource: InvalidFakeData
+
+    @BeforeEach
+    fun setup() {
+        fakeDataSource = CheapestApartmentsInCitiesFakeData()
+        emptyDataSource = EmptyFakeData()
```
Comment (msalarab) [inline src/test/kotlin/interactor/GetTopCitiesHasCheapestPriceOfApartmentsAndCountOfYearToBuyTest.kt]: change this class to object

**Review by msalarab — CHANGES_REQUESTED** [review]

**Review by msalarab — APPROVED** [review]

**Review by AmeerAmjed — APPROVED** [review]

**Pull request merged**